libyul/optimiser/SimplificationRules.cpp

```diff
@@ -171,12 +171,26 @@ bool Pattern::matches(
 			return false;
 		assertThrow(m_arguments.size() == instrAndArgs->second->size(), OptimizerException, "");
 		for (size_t i = 0; i < m_arguments.size(); ++i)
-			if (!m_arguments[i].matches(instrAndArgs->second->at(i), _dialect, _ssaValues))
+		{
+			Expression const& arg = instrAndArgs->second->at(i);
+			// If this is a direct function call instead of a variable or literal,
+			// we reject the match because side-effects could prevent us from
+			// arbitrarily modifying the code.
+			if (
+				holds_alternative<FunctionCall>(arg) ||
+				!m_arguments[i].matches(arg, _dialect, _ssaValues)
+			)
 				return false;
+		}
 	}
 	else
 	{
 		assertThrow(m_arguments.empty(), OptimizerException, "\"Any\" should not have arguments.");
+		// If this is a direct function call instead of a variable or literal,
+		// we reject the match because side-effects could prevent us from
+		// arbitrarily modifying the code.
+		if (holds_alternative<FunctionCall>(*expr))
```

Comment:
I was wondering if we have a pattern where this would be true. Perhaps it is possible to convert this into an assert.

Reply:
I converted this into `solAssert(!holds_alternative<FunctionCall>(*expr), "wrong");` and ran isoltest. There were no throws. Maybe we should change it into an assert.

Reply:
I'm pretty sure this can happen if you don't run the ExpressionSplitter before the ExpressionSimplifier.

Reply:
Even without `ExpressionSplitter` not sure if it can happen. If the outermost pattern is a built-in (i.e., the pattern type is an operation), then in line 180 each argument that is a function call is omitted from calling `Pattern::matches`.

Also, to get into this `if` branch, `m_kind` should be `Pattern::Any`, so if `expr` is a SSA Identifier, it won't be replaced by a functioncall.

So I think the only way to get here is if have a rule such as `{X, [=]{ return X; }}` but we don't have such rules.